### PR TITLE
feat(compiler): allow to use the grave accent character outside code blocks and add language inference from file extension

### PR DESCRIPTION
**Pull request by EmileRolley**
As mentioned in #120, the grave accent character ('`') isn't usable inside law text. So, this PR adds the possibility to use them outside code blocks.

### Changelog

* Add language inference in `driver.ml` based on the input file extension.
* Modify lexers in order to manage grave accent outside code blocks.
* Add the `test_literate/` folder with tests corresponding to the addition of the grave accent support.
* Modify the `%.out` rule of `tests/Makefile` in order to manage different language.

### CODE REVIEW

**Comment by denismerigoux** [comment]
This is amazing, thanks Émile ! I knew this limitation could be overcome with some regex magic but I wasn't sure of the invocation. I would gladly accept the PR, please extend your finding to the other lexers :)

**Comment by denismerigoux** [comment]
And do keep the literate programming test file :)

**Comment by EmileRolley** [comment]
> This is amazing, thanks Émile ! I knew this limitation could be overcome with some regex magic but I wasn't sure of the invocation. I would gladly accept the PR, please extend your finding to the other lexers :)

You're welcome! I will edit the lexers and try to do the same for the # and > characters. By the way, if anyone knows how to specify the start-of-line character with Sedlex, I would be interested :innocent:. However, I'm currently in the middle of my exams, so I'll pick this mr up at the end of these.

**Comment by EmileRolley** [comment]
> And do keep the literate programming test file :)

Alright, but it doesn't really test the correct formation of the AST...

**Comment by denismerigoux** [comment]
> However, I'm currently in the middle of my exams, so I'll pick this mr up at the end of these.

Yeah take your time, no rush here, your exams are more important than this :)

> You're welcome! I will edit the lexers and try to do the same for the # and > characters. By the way, if anyone knows how to specify the start-of-line character with Sedlex, I would be interested. 

I'm afraid it will be more complicated to remove the limitation for `#` and `>`, precisely because the way we lex the Catala source file does not make a distinction between lines of text. To be able to match  `#` and `>` at the beginning of the line, you would have to put `\n` as a stop character for the `LawText` regexp, which means storing each line of text separately before ultimately piecing them back together to form a paragraph.

> Alright, but it doesn't really test the correct formation of the AST...

That's fine, the fact that the compiler does not throw any error when parsing the file is good enough as a test.

**Comment by EmileRolley** [comment]
> I'm afraid it will be more complicated to remove the limitation for # and >, precisely because the way we lex the Catala source file does not make a distinction between lines of text. To be able to match # and > at the beginning of the line, you would have to put \n as a stop character for the LawText regexp, which means storing each line of text separately before ultimately piecing them back together to form a paragraph.
>

Is it desirable to add this ?

File: src/catala/driver.ml [hunk 1]
```diff
@@ -28,19 +36,29 @@ let driver (source_file : Pos.input_file) (debug : bool) (dcalc : bool) (unstyle
...
-      | Some l ->
-          if l = "fr" then `Fr
-          else if l = "en" then `En
-          else if l = "pl" then `Pl
-          else if l = "non-verbose" then `NonVerbose
-          else
+      | Some l -> l
+      | None -> (
+          (* Try to infer the language from the intput file extension. *)
+          let exts = List.rev (String.split_on_char '.' !filename) in
```
Comment (denismerigoux) [inline src/catala/driver.ml:47]: Use `Filename.extension` instead https://ocaml.org/api/Filename.html

File: src/catala/surface/lexer.ml [hunk 2]
```diff
@@ -554,7 +559,9 @@ let lex_law (lexbuf : lexbuf) : token =
         LAW_INCLUDE (Ast.PdfFile ((name, Pos.from_lpos pos), pages))
       else LAW_INCLUDE (Ast.CatalaFile (name, Pos.from_lpos pos))
   | Plus '#', Star white_space, Plus (Compl '\n'), Star white_space, '\n' -> get_law_heading lexbuf
-  | Plus (Compl ('#' | '`' | '>')) -> LAW_TEXT (Utf8.lexeme lexbuf)
+  | Plus (Compl ('#' | '`' | '>') | Rep ('`', 1 .. 2), Compl '`' | "```", (Plus white_space | '\n'))
```
Comment (denismerigoux) [inline src/catala/surface/lexer.ml:562]: A comment that describes the intent of this conjunction of regexes would be appreciated :)

File: src/catala/surface/lexer_en.ml [hunk 3]
```diff
@@ -515,7 +519,9 @@ let lex_law (lexbuf : lexbuf) : token =
         LAW_INCLUDE (Ast.PdfFile ((name, Pos.from_lpos pos), pages))
       else LAW_INCLUDE (Ast.CatalaFile (name, Pos.from_lpos pos))
   | Plus '#', Star white_space, Plus (Compl '\n'), Star white_space, '\n' -> get_law_heading lexbuf
-  | Plus (Compl ('#' | '`' | '>')) -> LAW_TEXT (Utf8.lexeme lexbuf)
+  | Plus (Compl ('#' | '`' | '>') | Rep ('`', 1 .. 2), Compl '`' | "```", (Plus white_space | '\n'))
```
Comment (denismerigoux) [inline src/catala/surface/lexer_en.ml:522]: Ditto for a comment

File: src/catala/surface/lexer_fr.ml [hunk 4]
```diff
@@ -524,7 +526,9 @@ let lex_law (lexbuf : lexbuf) : token =
         LAW_INCLUDE (Ast.PdfFile ((name, Pos.from_lpos pos), pages))
       else LAW_INCLUDE (Ast.CatalaFile (name, Pos.from_lpos pos))
   | Plus '#', Star white_space, Plus (Compl '\n'), Star white_space, '\n' -> get_law_heading lexbuf
-  | Plus (Compl ('#' | '`' | '>')) -> LAW_TEXT (Utf8.lexeme lexbuf)
+  | Plus (Compl ('#' | '`' | '>') | Rep ('`', 1 .. 2), Compl '`' | "```", (Plus white_space | '\n'))
```
Comment (denismerigoux) [inline src/catala/surface/lexer_fr.ml:529]: Ditto for a comment

File: src/catala/surface/lexer_pl.ml [hunk 5]
```diff
@@ -519,7 +527,9 @@ let lex_law (lexbuf : lexbuf) : token =
         LAW_INCLUDE (Ast.PdfFile ((name, Pos.from_lpos pos), pages))
       else LAW_INCLUDE (Ast.CatalaFile (name, Pos.from_lpos pos))
   | Plus '#', Star white_space, Plus (Compl '\n'), Star white_space, '\n' -> get_law_heading lexbuf
-  | Plus (Compl ('#' | '`' | '>')) -> LAW_TEXT (Utf8.lexeme lexbuf)
+  | Plus (Compl ('#' | '`' | '>') | Rep ('`', 1 .. 2), Compl '`' | "```", (Plus white_space | '\n'))
```
Comment (denismerigoux) [inline src/catala/surface/lexer_pl.ml:530]: Ditto comment

**Review by denismerigoux — COMMENTED** [review]
Thank you very much Émile ! Just a couple comments to explain why it works and we're good.

**Comment by EmileRolley** [comment]
> Thank you very much Émile ! Just a couple comments to explain why it works and we're good.

Here or in the code ?

**Comment by denismerigoux** [comment]
In the code, see my review comments

**Review by denismerigoux — APPROVED** [review]

**Pull request merged**